heatmap/src/heatmap.rs

```diff
@@ -280,52 +308,43 @@ impl Heatmap {
                 // a thread panic while locked.
                 if let Some(_lock) = self.lock.try_lock() {
                     // now that we have the lock, check that we still need to
-                    // tick forward
-                    if time < self.next_tick.load(Ordering::Relaxed) {
+                    // tick forward.
+                    if now < self.next_tick.load(Ordering::Relaxed) {
+                        // someone already finished tick maintenance
                         return;
                     }
 
-                    // calculate the index of the next current slice
-                    let current = self.current.load(Ordering::Relaxed);
-                    let mut next = current + 1;
-                    if next >= self.slices.len() {
-                        next -= self.slices.len();
-                    }
+                    let mut curr_tick = self.curr_tick.load(Ordering::Relaxed);
 
-                    // move current and next_tick forward
-                    self.current.store(next, Ordering::Relaxed);
-                    self.next_tick.fetch_add(self.resolution, Ordering::Relaxed);
+                    // calculate the number of histogram histogram slices that need cleanup
+                    let elapsed = now.checked_duration_since(curr_tick).unwrap();
+                    let mut ticks_forward = elapsed.as_nanos() / self.resolution.as_nanos();
 
-                    // now we have a slice to subtract and clear from the summary
-                    // this is the histogram that is one ahead of our new current
-                    // position
-                    let mut to_clear = next + 1;
-
-                    // check if we need to wrap around to the start
-                    if to_clear >= self.slices.len() {
-                        to_clear -= self.slices.len();
+                    // move current and next_tick forward
+                    curr_tick += Duration::from_nanos(self.resolution.as_nanos() * ticks_forward);
+                    next_tick = curr_tick + self.resolution;
+
+                    self.curr_tick.store(curr_tick, Ordering::Relaxed);
+                    self.next_tick.store(next_tick, Ordering::Relaxed);
+
+                    // Note that in steady state, the next slice of the previous "current"
+                    // histogram (which was marked by `idx`) is the oldest slice of histogram
+                    // and therefore needs to be cleaned up. Similarly, if we need to move
+                    // the tick forward by more than one window, all histograms we encounter
+                    // while advancing the `idx` will need to be cleaned up. Of course, here
+                    // the index will be wrapped around the ring buffer if it gets to the end.
+                    let mut idx = self.idx.load(Ordering::Relaxed) as usize;
+                    while ticks_forward > 0 {
```

Comment:
I don't think there's a downside here... but you could do `for _ in 0..ticks_forward {` instead and not have to decrement. `ticks_forward` no longer needs to be mutable then.

Reply:
yep, my C habit is showing, fixed.